### PR TITLE
Bump lower Dart SDK constraints to 3.0 & add class modifiers

### PR DESCRIPTION
**Pull request by goderbauer**
Part of https://github.com/flutter/flutter/issues/118837.

This will allow us to use the latest Dart 3.0 features.

This PR also adds class modifiers where necessary: 

* to all classes identified as mixin-able as part of the analysis in http://flutter.dev/go/mixin-modifier
* subclasses of LinkedListEntry (a newly marked `base` class) to make the analyzer warnings below happy
* slight adjustments to how `TapRegionRegistry` is used to fix analyzer warnings

```
  error • The type '_OverlayEntryLocation' must be 'base', 'final' or 'sealed' because the supertype 'LinkedListEntry' is 'base' • packages/flutter/lib/src/widgets/overlay.dart:1641:7 • subtype_of_base_or_final_is_not_base_final_or_sealed
  error • The type '_RenderDeferredLayoutBox' must be 'base', 'final' or 'sealed' because the supertype 'LinkedListEntry' is 'base' • packages/flutter/lib/src/widgets/overlay.dart:1957:7 • subtype_of_base_or_final_is_not_base_final_or_sealed
  error • The type '_ListenerEntry' must be 'base', 'final' or 'sealed' because the supertype 'LinkedListEntry' is 'base' • packages/flutter/lib/src/widgets/scroll_notification_observer.dart:38:7 • subtype_of_base_or_final_is_not_base_final_or_sealed
```

Blocked on waiting for https://github.com/flutter/engine/pull/40178 to roll into the framework.

### CODE REVIEW

**Comment by goderbauer** [comment]
FYI @jakemac53 @chingjun

**Comment by goderbauer** [comment]
@LongCatIsLooong Do the changes to `overlay.dart` look good to you?

@gspencergoog Do the changes in `tap_region.dart` look good to you to keep TapRegionRegistry as an interface instead of making it mixin-able?

**Review by LongCatIsLooong — APPROVED** [review]
YES! The overlay.dart changes LGTM. Pattern matching and record are still behind the experimental flag right?

**Comment by goderbauer** [comment]
> Pattern matching and record are still behind the experimental flag right?

Once this PR lands we're good to use them in the framework. This PR is still blocked on waiting for Dart v3.0.0-325.0.dev to roll into the engine & framework, though. That Dart version flips the flag on these features.

**Comment by jakemac53** [comment]
cc @johnniwinther it sounds like we should hold off on landing this until exhaustiveness checking is finished? What is the status and is there an issue we can track?

**Comment by johnniwinther** [comment]
@goderbauer Note that the errors you see currently are from a temporary exhaustiveness algorithm based on flow-analysis. The real algorithm, which is in the process of being enabled, will yield different results.

**Comment by goderbauer** [comment]
Full exhaustiveness checking is rolling into the framework with https://github.com/flutter/flutter/pull/123059.

**Review by christopherfujino — APPROVED** [review]
LGTM

**Comment by goderbauer** [comment]
To my surprise, with the bump to 3.0 the analyzer complained about a handful of additional places where `const` should be used, I fixed those up in the last commit.

**Comment by goderbauer** [comment]
Looks like I forgot to update some analyzer dependencies to a class-modifier compatible version. Working on that now...

**Comment by goderbauer** [comment]
There appears to be another issue with running tests on the web. I tried to narrow it down a little bit and filed https://github.com/dart-lang/sdk/issues/51802. This one is a blocker because it is making all the web tests fail and I am not aware of a work around.

**Comment by goderbauer** [comment]
I have a work around implemented for that issue (see commit named "work around for continue-switch-statement"), which basically removes the continue statement with a label from our code base. That statement appears to be the root cause of https://github.com/dart-lang/sdk/issues/51802.

@Piinks Could you review the commit named "work around for continue-switch-statement"? Thanks!

**Comment by goderbauer** [comment]
Google Testing is failing because the engine roll in https://github.com/flutter/flutter/pull/123059 has not rolled into Google yet.

File: packages/flutter/lib/src/material/app.dart
Comment (Piinks) [inline packages/flutter/lib/src/material/app.dart:1]: ScrollBehavior changed re: https://github.com/dart-lang/sdk/issues/51802 LGTM. Thank you!

File: packages/flutter/test/material/expansion_tile_test.dart
Comment (gspencergoog) [inline packages/flutter/test/material/expansion_tile_test.dart:334]: Are these going to show up in any customer code that uses prefer_const_constructors?  This lint is turned on as part of `flutter_lints`, so it might affect a lot of people if so.

Comment (goderbauer) [reply packages/flutter/test/material/expansion_tile_test.dart:334]: Yeah, they will see this once they bump the sdk constraints to 3.0 in their pubspecs. I am pushing hard on https://github.com/dart-lang/sdk/issues/51800 to get this fixed before Dart 3 is released to stable.

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

File: packages/flutter/test/material/expansion_tile_test.dart
Comment (gspencergoog) [reply packages/flutter/test/material/expansion_tile_test.dart:334]: Yeah, that seems important.

**Pull request merged**
